metaflow/plugins/kfp/kfp.py

```diff
@@ -242,80 +266,170 @@ def trigger(cls, kubernetes_namespace: str, name: str, parameters=None):
     ) -> Dict[str, Any]:
         """
         Creates a new Argo Workflow pipeline YAML using `kfp.compiler.Compiler()`.
         Note: Intermediate pipeline YAML is saved at `pipeline_file_path`
         """
-        pipeline_func, pipeline_conf = self.create_kfp_pipeline_from_flow_graph()
+        pipeline_func, pipeline_conf = self.create_kfp_pipeline_from_flow_graph(
+            flow_parameters
+        )
         workflow: Dict[Text, Any] = kfp.compiler.Compiler()._create_workflow(
             pipeline_func=pipeline_func,
             pipeline_conf=pipeline_conf,
         )
 
         workflow["spec"]["arguments"]["parameters"] = [
-            {
-                "name": "flow_parameters_json",
-                "value": json.dumps(flow_parameters if flow_parameters else {}),
-            }
+            dict(name=k, value=json.dumps(v) if isinstance(v, dict) else v)
+            for k, v in flow_parameters.items()
         ]
 
-        if output_format == "argo-workflow":
+        if kind == "Workflow":
             # Output of KFP compiler already has workflow["kind"] = "Workflow".
 
             # Keep generateName - Argo Workflow is usually used in single run.
 
             # Service account is added through webhooks.
             workflow["spec"].pop("serviceAccountName", None)
-        elif output_format == "argo-workflow-template":
+        elif kind == "WorkflowTemplate":
             workflow["kind"] = "WorkflowTemplate"
+            workflow["spec"]["serviceAccountName"] = (
+                KUBERNETES_SERVICE_ACCOUNT or "default-editor"
+            )
 
             # Use static name to make referencing easier.
             # Note the name has to follow k8s format.
             # self.name is typically CamelCase as it's python class name.
             # generateName contains a sanitized version of self.name from kfp.compiler
-            workflow["metadata"]["name"] = (
+            workflow_name = (
                 name if name else workflow["metadata"].pop("generateName").rstrip("-")
             )
+            workflow["metadata"]["name"] = workflow_name
 
             # Service account is added through webhooks.
             workflow["spec"].pop("serviceAccountName", None)
         else:
-            raise NotImplementedError(f"Unsupported output format {output_format}.")
+            raise NotImplementedError(f"Unsupported output format {kind}.")
+
+        if max_run_concurrency and max_run_concurrency > 0:
+            workflow["spec"]["synchronization"] = {
+                "semaphore": {
+                    "configMapKeyRef": {
+                        "name": sanitize_k8s_name(self.name),
+                        "key": "max_run_concurrency",
+                    }
+                }
+            }
 
         return workflow
 
-    def create_run_on_argo(
-        self, kubernetes_namespace: str, flow_parameters: dict
+    @staticmethod
+    def _config_map(workflow_name: str, max_run_concurrency: int):
```

Comment:
It looks like `max_run_concurrency` is an `Optional[int]` elsewhere. This function needs to guard against `None` value, or we drop the `Optional` in all other places.

Reply:
fixed!